kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/PartitionUtils.java

```diff
@@ -100,6 +105,108 @@ public static ColumnarBatch withPartitionColumns(
+            return new Tuple2(ALWAYS_TRUE, predicate);
+        } else {
+            return new Tuple2<>(predicate, ALWAYS_TRUE);
+        }
+    }
+
+    /**
+     * Utility method to rewrite the partition predicate referring to the table schema as predicate
+     * referring to the {@code partitionValues} in scan files read from Delta log. The scan file
+     * batch is returned by the {@link io.delta.kernel.Scan#getScanFiles(TableClient)}.
+     *
+     * @param predicate             Predicate containing filters only on partition columns.
+     * @param partitionColNameTypes Map of partition columns and their types.
+     * @return
+     */
+    public static Predicate rewritePartitionPredicateOnScanFileSchema(
+        Predicate predicate, Map<String, DataType> partitionColNameTypes) {
+        return new Predicate(
+            predicate.getName(),
+            predicate.getChildren().stream()
+                .map(child -> rewritePartitionColumnRef(child, partitionColNameTypes))
+                .collect(Collectors.toList()));
+    }
+
+    private static Expression rewritePartitionColumnRef(
+        Expression expression, Map<String, DataType> partitionColNameTypes) {
+        Column scanFilePartitionValuesRef = InternalScanFileUtils.ADD_FILE_PARTITION_COL_REF;
+        if (expression instanceof Column) {
+            Column column = (Column) expression;
+            String partColName = column.getNames()[0];
+            DataType partColType = partitionColNameTypes.get(partColName);
+
+            Expression elementAt =
+                new ScalarExpression(
+                    "element_at",
+                    asList(scanFilePartitionValuesRef, Literal.ofString(partColName)));
+
+            if (partColType instanceof StringType) {
+                return elementAt;
+            }
+
+            // Add expression to decode the partition value based on the partition column type.
+            return new PartitionValueExpression(elementAt, partColType);
+        } else if (expression instanceof Predicate) {
+            return rewritePartitionPredicateOnScanFileSchema(
+                (Predicate) expression, partitionColNameTypes);
+        }
+
+        return expression;
+    }
+
+    private static boolean hasNonPartitionColumns(
+        List<Expression> children,
+        Set<String> partitionColNames) {
+        for (Expression child : children) {
+            if (child instanceof Column) {
+                String[] names = ((Column) child).getNames();
+                if (names.length != 1 || !partitionColNames.contains(names[0])) {
+                    return true;
+                }
+            } else {
+                return hasNonPartitionColumns(child.getChildren(), partitionColNames);
+            }
+        }
+        return false;
+    }
+
+    private static Predicate combineWithAndOp(Predicate left, Predicate right) {
+        if (left.getName().equalsIgnoreCase("ALWAYS_TRUE")) {
```

Comment:
Currently, with input: `left=ALWAYS_FALSE,right=$predicate` we are returning `AND(ALWAYS_FALSE, $predicate)`

but since one of the inputs is `ALWAYS_FALSE`, and we are performing an `AND` operation, we can just short circuit early and return ALWAYS_FALSE, no?

Reply:
The splitting utility method, should never returns `ALWAYS_FALSE`, but can add the handling here incase that changes in futrue.

Reply:
Addressed.